packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts

```diff
@@ -622,13 +617,30 @@ export class OdspDocumentStorageService implements IDocumentStorageService {
             timeout: hostSnapshotOptions?.timeout ? Math.min(hostSnapshotOptions.timeout, this.maxSnapshotFetchTimeout) : this.maxSnapshotFetchTimeout,
         };
 
-        // No limit on size of snapshot, as otherwise we fail all clients to summarize
+        // No limit on size of snapshot or time to fetch, as otherwise we fail all clients to summarize
         if (this.hostPolicy.summarizerClient) {
             snapshotOptions.mds = undefined;
+            snapshotOptions.timeout = undefined;
         }
 
+        const snapshotDownloader = async (url: string, fetchOptions: {[index: string]: any}) => {
+            return this.epochTracker.fetchAndParseAsJSON<IOdspSnapshot>(
+                url,
+                fetchOptions,
+                "treesLatest",
+                true,
+            );
+        };
+
         try {
-            const odspSnapshot = await this.fetchSnapshotCore(snapshotOptions, tokenFetchOptions);
+            const { odspSnapshot } = await fetchLatestSnapshotCore(
+                this.odspResolvedUrl,
+                this.getStorageToken,
+                tokenFetchOptions,
+                snapshotOptions,
+                this.logger,
+                snapshotDownloader,
+                this.cache.persistedCache);
```

Comment:
Something does not add up here. this.cache.persistedCache is of IPersistedFileCache type. But fetchLatestSnapshotCore takes IPersistedCache. It's not clear why compiler is happy (is it?)

The problem as it's now (and that's my fault, sorry, I suggested to move into this direciton), is that we will wrap the value (snapshot) twice with IVersionedValueWithEpoch - once in side fetchLatestSnapshotCore, and once inside of EpochTracker (which implements this.cache.persistedCache).

I think the best solution is to use IPersistedFileCache in fetchLatestSnapshotCore, and have prefetchLatestSnapshot wrap IPersistedCache to produce IPersistedFileCache. Or even simplify further and just pass custom callback that stores value in cache. Latter is cleaner, as that's the only dependency of this function, it does not need to know anything else (and actual construction of entry can be outside of this function).

I think overall it begs a question - how do we find bugs IF for whatever reason the key that we use to start data in prefetch and normal load flows are different or formats are different that we can't understand it and silently discard it?

Reply:
Please double check me, it's a bit weird that it compiles... (or maybe I got it all wrong?)

Reply:
So, I added callback to cache the snapshot. That callback will put the appropriate value in cache.
For key that we use to get and put values from cache, I have added a function in utils, so that it remain consistent across calls and places.

Reply:
Also removed the promise in promise :)